GitCommands/RevisionReader.cs

```diff
@@ -511,34 +466,86 @@ public StringLineReader(string s)
-                {
-                    _index = endIndex + 1;
-                }
+                _index = endIndex + 1;
 
                 return pool is not null
                     ? pool.Intern(_s, startIndex, endIndex - startIndex)
                     : _s.Substring(startIndex, endIndex - startIndex);
             }
 
-            public string? ReadToEnd(bool advance = true)
+            public StringLineReader TrimStart()
+            {
+                while (_index < _s.Length && char.IsWhiteSpace(_s[_index]))
+                {
+                    _index++;
+                }
+
+                return this;
+            }
+
+            private void TrimEnd(ref int endIndex)
+            {
+                while (endIndex > _index && char.IsWhiteSpace(_s[endIndex - 1]))
+                {
+                    endIndex--;
+                }
+            }
+
+            public (string? subject, string? body, bool hasMultiLineMessage) ReadToEndOfBody(bool skipBody = false)
             {
                 if (_index == _s.Length)
                 {
-                    return null;
+                    return (null, null, false);
+                }
+
+                int indexOfEndOfBody = _s.LastIndexOf(EndOfBody, _s.Length - 1, _s.Length - _index);
+                if (indexOfEndOfBody < 0)
+                {
+                    return (null, null, false);
+                }
+
+                // Trim end
+                int endBodyIndex = indexOfEndOfBody;
+                TrimEnd(ref endBodyIndex);
+
+                int endSubjectIndex = _s.IndexOf('\n', _index, endBodyIndex - _index);
+                if (endSubjectIndex < 0)
+                {
+                    endSubjectIndex = endBodyIndex;
+                }
+                else
+                {
+                    TrimEnd(ref endSubjectIndex);
                 }
 
-                var s = _s.Substring(_index);
+                // Subject can also be defined as the contents before empty line (%s for --pretty),
+                // this uses the alternative definition of first line in body.
+                string subject = _s.Substring(_index, endSubjectIndex - _index);
+                bool hasMultiLineMessage = endSubjectIndex < endBodyIndex;
+
+                // Skip body if identical to subject
+                string? body = skipBody || !hasMultiLineMessage
+                    ? null
+                    : _s.Substring(_index, endBodyIndex - _index);
+
+                _index = 1 /* EndOfBody.Length */ + indexOfEndOfBody;
+
+                return (subject, body, hasMultiLineMessage);
+            }
 
-                if (advance)
+            public string? ReadToEnd()
+            {
+                if (_index >= _s.Length)
                 {
-                    _index = _s.Length;
+                    return null;
                 }
 
-                return s;
+                _index = _s.Length;
+
+                return _s.Substring(_index);
```

Comment:
The substring must be taken before setting `_index = _s.Length`. The latter must be done for the `return null`, too.

Reply:
One reason I removed the advance of index here.
Renaming to `PeakToEnd()`, reversing the advancing

Reply:
I agree, but `PeekToEnd`.